backend/OKPluginVariableRendering/VariableRendering.cs

```diff
@@ -0,0 +1,358 @@
+
+                                    if (tmpCIAttributes.ContainsKey(a.Name))
+                                    {
+                                        // check if current attribute has higher priority
+                                        if (tmpCIAttributes[a.Name].Priority < a.Priority)
+                                        {
+                                            tmpCIAttributes[a.Name] = a;
+                                        }
+                                    } else
+                                    {
+                                        tmpCIAttributes.Add(a.Name, a);
+                                    }
+
+                                    attribute.Name = a.Name;
+                                    attribute.Value = a.Value;
+                                }
+                            }
+                        }
+
+                        gatheredAttributes.AddRange(tmpCIAttributes.Values);
+                    }
+                }
+
+                // for all gathered attributes insert them to the main ci
+                foreach (var mapping in cfg.BaseCI.AttributeMapping)
+                {
+
+                    foreach (var attribute in gatheredAttributes)
+                    {
+                        // first we need to check input whitelist and blacklist
+                        if (!IsAttributeAllowed(attribute.Name, cfg.BaseCI.InputWhitelist, cfg.BaseCI.InputBlacklist))
+                        {
+                            continue;
+                        }
+
+                        // check base ci attribute mapping
+                        if (attribute.Name == "__name")
+                        {
+
+                        }
+                        else if(!IsAttributeIncludedInSource(attribute.Name, mapping.Source))
+                        {
+                            continue;
+                        }
+
+                        // check if attribute exists in base ci
+                        if (mainCI.MergedAttributes.ContainsKey(attribute.Name) && ciPrio < attribute.Priority)
+                        {
+                            // don't change this attribute
+                            continue;
+                        }
+
+                        fragments.Add(new BulkCIAttributeDataLayerScope.Fragment(GetTargetName(attribute.Name, mapping.Target), attribute.Value, mainCI.ID));
+
+                    }
+
+                }
+            }
+
+            await attributeModel.BulkReplaceAttributes(
+                new BulkCIAttributeDataLayerScope("", targetLayer.ID, fragments),
+                changesetProxy, 
+                new DataOriginV1(DataOriginType.ComputeLayer), 
+                trans,
+                MaskHandlingForRemovalApplyNoMask.Instance);
+
+            return true;
+        }
+
+
+        private static bool IsAttributeAllowed(string attribute, List<string> attributeWhitelist, List<string> attributeBlacklist)
```

Comment:
This seems pretty buggy. I tested this with the following inputs and it returned `false`:
attribute = monitoring.vars.cmdb.interface.ip
attributeWhitelist = *
attributeBlacklist = monitoring.vars.do_not_include.*

That can't be right. I believe it would be best to write some unit tests for this.

Reply:
I checked this and found that the bug was with the range operator the substring was selected incorrectly. For this I have added some test and tested some scenarios.